### PR TITLE
Add Whippet to bottlerocket

### PR DESCRIPTION
**Pull request by yeazelm**
**Issue number:** [ 2449](https://github.com/bottlerocket-os/bottlerocket/issues/2449)

**Description of changes:**
This is an initial commit for whippet. There is more work to be done to make this robust, but this glues together the bits so that DNS works on top of https://github.com/bottlerocket-os/bottlerocket/pull/3266. With both of these changes, we should have a basic working stack. 

This relies on systemd to restart `whippet` often if it fails. This works well enough for now, but the overall approach needs to be hardened up for a proper daemon that we want for this purpose.

Things this doesn't do:
* Handle errors nicely, much better error handling will need to be done to make `whippet` not die all the time. 
* Creates a proper loop for waiting on D-Bus signals. Instead if fires off the `netdog` command and relies upon restart by systemd
* Nice signal handling. It will die if restarted by `systemctl` and more or less is functional, but there are likely edge cases that would make this less robust for production
* Have proper logging with levels. Right now it just uses `println!()`
* Have unit or integration tests

There will be more work to come to make this better, but I wanted to get a starting point out in the wild for folks to look at.

I confirmed that when enabling the image-feature of systemd-networkd, whippet shows up in the image:
```bash
find /mnt -name "*whippet*"
/mnt/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/whippet.service
/mnt/x86_64-bottlerocket-linux-gnu/sys-root/usr/bin/whippet
```
and otherwise whippet is excluded from the image.

**Testing done:**
I have done a lot of manual testing on QEMU and EC2. I have basic output from my EC2 instance:
```bash
bash-5.1# ping amazon.com
PING amazon.com (205.251.242.103) 56(84) bytes of data.
64 bytes from s3-console-us-standard.console.aws.amazon.com (205.251.242.103): icmp_seq=1 ttl=219 time=64.1 ms
64 bytes from s3-console-us-standard.console.aws.amazon.com (205.251.242.103): icmp_seq=2 ttl=219 time=63.2 ms

--- amazon.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1002ms
rtt min/avg/max/mdev = 63.186/63.645/64.105/0.459 ms

bash-5.1# journalctl -u whippet.service
Jul 19 02:17:46 localhost systemd[1]: Starting Bottlerocket D-Bus listener for network events...
Jul 19 02:17:46 localhost systemd[1]: Started Bottlerocket D-Bus listener for network events.
Jul 19 02:17:46 localhost whippet[1042]: Calling netdog to get primary interface
Jul 19 02:17:46 localhost whippet[1042]: Primary interface is eth0
Jul 19 02:17:46 localhost whippet[1042]: Link id: 2 Name: eth0 Path: OwnedObjectPath(ObjectPath("/org/freedesktop/network1/link/_32"))
Jul 19 02:17:46 localhost whippet[1042]: Found eth0 is configuring
Jul 19 02:17:46 localhost whippet[1042]: Link id: 1 Name: lo Path: OwnedObjectPath(ObjectPath("/org/freedesktop/network1/link/_31"))
Jul 19 02:17:46 localhost whippet[1042]: DEBUG: found lo but is not eth0
Jul 19 02:17:48 localhost whippet[1042]: org.freedesktop.network1.Link.AdministrativeState changed to `Str(Str(Borrowed("configured")))`
Jul 19 02:17:48 localhost whippet[1042]: org.freedesktop.network1.Link.IPv6AddressState changed to `Str(Str(Borrowed("degraded")))`
```
The output is rough for now and could be cleaned up.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: packages/os/os.spec [hunk 1]
```diff
@@ -346,6 +353,7 @@ echo "** Output from non-static builds:"
     -p certdog \
     -p shimpei \
     -p bloodhound \
+    -p whippet \
```
Comment (zmrow) [inline packages/os/os.spec]: This is building whippet all the time, regardless of the network backend.  We could probably conditionalize it so it's not getting included in things before it's completed

File: sources/whippet/README.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# whippet
+
+Current version: 0.1.0
+
+whippet is a D-Bus listener that reponds to events on D-Bus.
+Currently, it is used to notify netdog of events on the network1 D-Bus service.
```
Comment (zmrow) [inline sources/whippet/README.md]: nit: Should probably read: `...of events for the primary interface on the network1 D-Bus service.`

File: sources/whippet/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+*/
+use futures::stream::StreamExt;
+use snafu::{ensure, ResultExt};
+use std::process;
+use std::process::Command;
+use std::string::String;
+use zbus::zvariant::OwnedObjectPath;
+use zbus::Connection;
+
+pub(crate) mod networkmanager;
```
Comment (zmrow) [inline sources/whippet/src/main.rs]: This probably doesn't need to be `pub` because it's not being shared to any other module. :)

File: sources/whippet/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        .output()
+        .context(error::NetdogExecutionSnafu)?;
+    ensure!(
+        primary_interface_name_result.status.success(),
+        error::FailedNetdogSnafu {
+            stderr: String::from_utf8_lossy(&primary_interface_name_result.stderr)
+        }
+    );
+
+    let clean_string = |s: String| s.trim().to_lowercase().trim_matches('"').to_string();
```
Comment (zmrow) [inline sources/whippet/src/main.rs]: Is the closure necessary here?  This logic is only being used in a single place.

File: sources/whippet/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    let conn = Connection::system().await?;
+
+    // Get the manager created from the connection to the system bus
+    let manager = NetworkManagerProxy::new(&conn).await?;
+
+    // List the lists on the bus
+    let links = manager.list_links().await?;
+
+    // Put the path in an option since we might not find it
+    let mut path_to_primary: Option<&OwnedObjectPath> = None;
```
Comment (zmrow) [inline sources/whippet/src/main.rs]: nit: I might put this variable closer to where it's used.  I sorta lost the context by the time it's used.

File: sources/whippet/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                        stderr: String::from_utf8_lossy(&primary_interface_status_result.stderr)
+                    }
+                );
+            }
+        } else {
+            println!("DEBUG: found {} but is not {}", name, &primary_interface);
+        }
+    }
+
+    // If we found the primary device, start listing to events
```
Comment (zmrow) [inline sources/whippet/src/main.rs]: s/listing/listening

File: sources/whippet/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            let args = signal.args()?;
+
+            for (name, value) in args.changed_properties().iter() {
+                println!(
+                    "{}.{} changed to `{:?}`",
+                    args.interface_name(),
+                    name,
+                    value
+                );
+                // Call netdog primary-interface to get the name
```
Comment (zmrow) [inline sources/whippet/src/main.rs]: I don't think we're getting the name here :)

File: sources/whippet/src/networkmanager.rs [hunk 8]
```diff
@@ -0,0 +1,144 @@
+use serde::Deserialize;
+use zbus::{dbus_proxy, Result};
+
+#[dbus_proxy(
+    default_service = "org.freedesktop.network1",
+    interface = "org.freedesktop.network1.Manager",
+    default_path = "/org/freedesktop/network1",
+    assume_defaults = false
+)]
+trait NetworkManager {
```
Comment (zmrow) [inline sources/whippet/src/networkmanager.rs:19]: I'm on the fence about including all of these.  I do appreciate the effort it took to gather this information and the fact we _could_ use them, but I think we should probably whittle this down to what we _do_ use.  I would like to save them for future reference and use, though...  Perhaps we comment them out with a message "for potential future use" or something?

**Review by zmrow — COMMENTED** [review]
Mostly nits since I know this is POC code.

The only thing I'd like us to investigate before merging is the fact the package is built for everything, not just variants using systemd-networkd.

File: sources/whippet/src/main.rs (hunk 4 above)
Comment (yeazelm) [reply sources/whippet/src/main.rs]: I lifted this logic from `netdog` and kept the shape but that makes sense now that I look at it. I'll just use it directly.

File: sources/whippet/src/networkmanager.rs (hunk 8 above)
Comment (yeazelm) [reply sources/whippet/src/networkmanager.rs:19]: I agree we should trim this down. These are autogenerated from `zbus_xmlgen` so we can always go back and get them again. I think I can trim these down as long as I document where we got them and how one would efficiently add things back in.

File: packages/os/os.spec (hunk 1 above)
Comment (yeazelm) [reply packages/os/os.spec]: I think I got the conditional building right now.

File: sources/whippet/src/networkmanager.rs (hunk 8 above)
Comment (yeazelm) [reply sources/whippet/src/networkmanager.rs:19]: I pulled out many of the things we are unlikely to use (the write stuff) but left some of the other read functions since its still early and I realized we might have nicer ways to do this. For example, we might shift to `get_link_by_name` vs `list_links`. I also anticipate us finding edge cases where the `*_state` calls might be needed. We can trim as we discover they are not needed as we test this code more.

**Review by zmrow — APPROVED** [review]
I'm good to merge this as POC code that won't be included in current variants.  I'll pick up the torch once it's merged and can continue the outstanding items.

**Review by stmcginnis — APPROVED** [review]
Looks good to me. And if we find anything wrong, we have some time to correct and adjust. Probably better to have this in and usable to find out if that is the case.

**Review by etungsten — APPROVED** [review]

File: packages/os/os.spec [hunk 9]
```diff
@@ -287,6 +289,11 @@ Summary: Compliance check framework
 %description -n %{_cross_os}bloodhound
 %{summary}.
 
+%package -n %{_cross_os}whippet
+Summary: D-Bus listener and marshaller
+%description -n %{_cross_os}whippet
+%{summary}.
+
```
Comment (bcressey) [inline packages/os/os.spec:296]: nit: you should make the whole sub-package conditional rather than creating an empty rpm with no files

File: packages/os/whippet.service [hunk 10]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Bottlerocket D-Bus listener for network events
```
Comment (bcressey) [inline packages/os/whippet.service:2]: nit: doesn't necessarily need to be Bottlerocket-specific

```suggestion
Description=D-Bus listener for network events
```

File: sources/whippet/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+authors = ["Matthew Yeazel <yeazelm@amazon.com>"]
+license = "Apache-2.0 OR MIT"
+edition = "2021"
+publish = false
+# Don't rebuild crate just because of changes to README.
+exclude = ["README.md"]
+
+[dependencies]
+snafu = "0.7"
+zbus = "3.8.0"
```
Comment (bcressey) [inline sources/whippet/Cargo.toml:13]: I'd rather not mix async runtimes:
```suggestion
zbus = { version = "3.8.0", default-features = false, features = ["tokio"] }
```

File: sources/deny.toml [hunk 12]
```diff
@@ -76,7 +76,10 @@ skip-tree = [
     { name = "windows-sys", version = "=0.42.0" },
     # generate-readme pulls in an older clap that causes some duplicate
     # dependencies
-    { name = "generate-readme", version = "=0.1.0" }
+    { name = "generate-readme", version = "=0.1.0" },
+    # gptman is locked to older nix but is fine to move up
+    # https://github.com/rust-disk-partition-management/gptman/pull/113
```
Comment (bcressey) [inline sources/deny.toml:81]: That pull request was closed and says it's not fine to move up.

File: sources/whippet/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,164 @@
+/*!
+whippet is a D-Bus listener that reponds to events on D-Bus.
```
Comment (bcressey) [inline sources/whippet/src/main.rs:2]:
```suggestion
whippet is a D-Bus listener that responds to events on D-Bus.
```

File: sources/whippet/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        .context(error::PrimaryInterfaceStringSnafu {})?;
+    let primary_interface: String = primary_interface_output_str
+        .trim()
+        .to_lowercase()
+        .trim_matches('"')
+        .to_string();
+    println!("Primary interface is {}", &primary_interface);
+
+    // Put the path in an option since we might not find it
+    let mut path_to_primary: Option<&OwnedObjectPath> = None;
```
Comment (bcressey) [inline sources/whippet/src/main.rs:52]: whippet should have a config file that tells it what interface(s) to watch and what commands to run in response.

File: sources/whippet/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+                let primary_interface_status_result = Command::new(NETDOG)
+                    .arg("write-primary-interface-status")
+                    .output()
+                    .context(error::NetdogExecutionSnafu)?;
+                ensure!(
+                    primary_interface_status_result.status.success(),
+                    error::FailedNetdogSnafu {
+                        stderr: String::from_utf8_lossy(&primary_interface_status_result.stderr)
+                    }
+                );
```
Comment (bcressey) [inline sources/whippet/src/main.rs:82]: I expect you'll want to use `tokio::process::Command` to avoid causing problems with the async runtime.

This seems like it should be a log message, either a warning or an error, and not cause the process to terminate.

File: sources/whippet/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        } else {
+            println!("DEBUG: found {} but is not {}", name, &primary_interface);
+        }
```
Comment (bcressey) [inline sources/whippet/src/main.rs:86]: This should all be standard logging statements - `debug!`, `warn!` etc.

File: sources/whippet/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+            println!(
+                "Found {} is {}",
+                &primary_interface, link_status.administrative_state
+            );
+            if link_status.administrative_state == "configured" {
+                // call netdog now since its already configured, then the polling for changes can block
+                println!(
+                    "Calling netdog write-primary-interface-status for {}",
+                    link_status.name
+                );
```
Comment (bcressey) [inline sources/whippet/src/main.rs:72]: This is racy because the interface may not be configured here, but may become configured before we are listening for the signal that tells us it's now configured.
